File: neetcode_150/longestRepeatingCharacterReplacement.java
//https://leetcode.com/problems/longest-repeating-character-replacement/

import java.util.HashMap;
import java.util.HashSet;

public class longestRepeatingCharacterReplacement {
    public static void main(String[] args) {

    }
    public int characterReplacement(String s, int k) {
        if(s.length()<=1){
            return s.length();
        }
        int left = 0, right = 0;
        int target = k;
        int answer = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        int top = 1;
        while(right<s.length()){
          int window = right-left+1;
          if(window-top<=k){
              right++;
              if(map.containsKey(s.charAt(right))){
                  map.put(s.charAt(right), map.get(s.charAt(right))+1);
                  top = Math.max(top, map.get(s.charAt(right)));
              }
              else{
                  map.put(s.charAt(right),1);
              }

          }
          else{
              map.put(s.charAt(left), map.get(s.charAt(left))-1);
              left++;
          }
          answer = Math.max(answer, window);
        }
        return answer;
    }
}

File: neetcode_150/longestRepeatingCharacterReplacement.py
def characterReplacement(self, s: str, k: int) -> int:
    freq = [0] * 26
    left = 0
    right = 0
    maxLength = 0
    maxFreq = 0
    while right < len(s):
        position = ord(s[right]) - ord('A')
        freq[position] += 1
        maxFreq = max(maxFreq, freq[position])

        while (right - left + 1) - maxFreq > k:
            p2 = ord(s[left]) - ord('A')
            freq[p2] -= 1
            left += 1
            maxFreq = max(freq)

        maxLength = max(maxLength, right-left+1)
        right+=1

    return maxLength